### PR TITLE
[WIP] Power manifold

### PR DESCRIPTION
**Pull request by mateuszbaran**
I've made an initial part of the power manifold. I've written `HybridArrays.jl` to make them faster but the current version is not benchmarked enough to be sure how much it helps but I'm optimistic about it.

### CODE REVIEW

**Comment by sdaxen** [comment]
Awesome!! I'm excited to look at this further. Should be able to review near the end of this week.

File: src/PowerManifold.jl [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+`TSize` statically defines the number of elements along each axis.
+For example, a manifold-valued time series would be represented by a power
+manifold with $n$ equal to 1 and $N_1$ equal to the number of samples.
+A manifold-valued image (for example in diffusion tensor imaging) would
+be represented by a two-axis power manifold ($n=2$) with $N_1$ and $N_2$
+equal to width and height of the image.
+
+While the size of the manifold is static, points on the power manifold
+would not be represented by statically-sized arrays. Operations on small
+power manifolds might be faster if they are represented as [`PowerManifold`](@ref).
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/PowerManifold.jl]: Yes, thanks.

File: src/PowerManifold.jl [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+# Constructor
+
+    PowerManifold(M, (N_1, N_2, ..., N_n))
+
+generates the power manifold $M^{N_1 \times N_2 \times \dots \times N_n}$.
+"""
+struct PowerManifold{TM<:Manifold, TSize} <: Manifold
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/PowerManifold.jl]: What shared functionality are you thinking about? I don't really see any right now.

By the way, thank you for the review :slightly_smiling_face: .

**Comment by mateuszbaran** [comment]
The PR is almost ready, there is just one minor change that needs to be done once HybridArrays 0.3.0 is tagged. The code assumes that the power manifold is represented by one big array (any kind should do but `HybridArray` is quite fast).

Do you think I should make any changes here @sethaxen @kellertuer ?

File: src/PowerManifold.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (kellertuer) [reply src/PowerManifold.jl]: I am not sure, whether a common base class is useful. They share that they build a manifold from another one/others, so I also don't see exactly what a common base class might bring as an advantage.

File: src/PowerManifold.jl [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    end
+    return radius
+end
+
+"""
+    PowerRetraction(retraction::AbstractRetractionMethod)
+
+Power retraction based on `retraction`. Works on [`PowerManifold`](@ref).
+"""
+struct PowerRetraction{TR<:AbstractRetractionMethod} <: AbstractRetractionMethod
```
Comment (kellertuer) [inline src/PowerManifold.jl]: Does this type just “power up” a retraction or can that also be a retraction that's different (though I can not imagine a case where I would use a different retraction for components 1 and 2)...

File: src/PowerManifold.jl [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+"""
+    PowerFVectorDistribution([type::VectorBundleFibers], [x], distr)
+
+Generates a random vector at point `x` from vector space (a fiber of a tangent
+bundle) of type `type` using the power distribution of `distr`.
+
+Vector space type and `x` can be automatically inferred from distribution `distr`.
+"""
+struct PowerFVectorDistribution{TSpace<:VectorBundleFibers{<:VectorSpaceType, <:PowerManifold}, TD<:FVectorDistribution, TX} <: FVectorDistribution{TSpace, TX}
```
Comment (kellertuer) [inline src/PowerManifold.jl]: What does the `F` in the name stand for? `Fiber`? Then one could also write that out.

File: benchmark/benchmarks.jl [hunk 5]
```diff
@@ -152,6 +153,69 @@ function add_manifold_benchmarks()
             test_tangent_vector_broadcasting = false)
     end
 
+    # power manifolds
+    begin
+        Ms = Sphere(2)
+        Ms1 = PowerManifold(Ms, (5,))
+        Ms2 = PowerManifold(Ms, (5,7))
+        Mr = Manifolds.Rotations(3)
+        Mr1 = PowerManifold(Mr, (5,))
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply benchmark/benchmarks.jl]: I think we can just have `PowerManifold(Mr, 5)` and `PowerManifold(Mr, 5, 7)`, right?

File: src/PowerManifold.jl (hunk 3 above)
Comment (mateuszbaran) [reply src/PowerManifold.jl]: This is just the "power up" of a single retraction. If someone needs a different retraction, making a new one is fairly simple.

File: src/PowerManifold.jl (hunk 4 above)
Comment (mateuszbaran) [reply src/PowerManifold.jl]: That's a reference to this type: https://github.com/JuliaNLSolvers/Manifolds.jl/blob/0ac128450df3e2501764a27b7955362dbd9925c8/src/VectorBundle.jl#L106 . Do you think it should be renamed to something like `FiberVector`?

File: benchmark/benchmarks.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (kellertuer) [reply benchmark/benchmarks.jl]: Yes, that would also be very nice, to have the last option you mentioned :)

File: src/PowerManifold.jl (hunk 3 above)
Comment (kellertuer) [reply src/PowerManifold.jl]: Ok, great.

File: src/PowerManifold.jl (hunk 4 above)
Comment (kellertuer) [reply src/PowerManifold.jl]: Ah, I see, so it's similar to `MPoint` and `TVector`? I didn't have that in mind. Then, of course, it'S fine to stay with that.

File: benchmark/benchmarks.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply benchmark/benchmarks.jl]: OK, I've changed that.

**Comment by mateuszbaran** [comment]
I'll merge it soon if there are no significant issues.

**Pull request merged**